File: packages/client/src/primordial/sandbox/primordial_delegate.mjs
/**
 * Primordial Delegation SDK for Node.js
 *
 * Provides a simple async API for agents to search, spawn, and interact with
 * other agents on the Primordial AgentStore via the delegation socket.
 *
 * Usage:
 *   import { search, runAgent, messageAgent, stopAgent } from './primordial_delegate.mjs';
 *
 * Zero dependencies — uses only Node built-ins (net).
 */

import { createConnection } from 'net';

const SOCK_PATH = '/tmp/_primordial_delegate.sock';

// ---------------------------------------------------------------------------
// Socket helpers (fresh connection per call)
// ---------------------------------------------------------------------------

function connect() {
  return new Promise((resolve, reject) => {
    const sock = createConnection(SOCK_PATH, () => resolve(sock));
    sock.on('error', reject);
  });
}

function send(sock, obj) {
  sock.write(JSON.stringify(obj) + '\n');
}

function readLines(sock) {
  let buf = '';
  const lines = [];
  let resolver = null;
  let done = false;

  sock.setEncoding('utf8');
  sock.on('data', (chunk) => {
    buf += chunk;
    let idx;
    while ((idx = buf.indexOf('\n')) !== -1) {
      const line = buf.slice(0, idx);
      buf = buf.slice(idx + 1);
      if (line.trim()) {
        const msg = JSON.parse(line);
        if (resolver) {
          const r = resolver;
          resolver = null;
          r(msg);
        } else {
          lines.push(msg);
        }
      }
    }
  });

  sock.on('end', () => {
    done = true;
    if (resolver) {
      const r = resolver;
      resolver = null;
      r(null);
    }
  });

  return {
    next() {
      if (lines.length > 0) return Promise.resolve(lines.shift());
      if (done) return Promise.resolve(null);
      return new Promise((resolve) => { resolver = resolve; });
    },
  };
}

async function request(msg) {
  const sock = await connect();
  try {
    send(sock, msg);
    const reader = readLines(sock);
    const result = await reader.next();
    if (result?.type === 'error') {
      throw new Error(result.error || 'unknown error');
    }
    return result;
  } finally {
    sock.destroy();
  }
}

async function* requestStream(msg) {
  const sock = await connect();
  const reader = readLines(sock);
  try {
    send(sock, msg);
    while (true) {
      const result = await reader.next();
      if (!result) return;
      yield result;
      if (result.type === 'error') return;
      if (result.done) return;
      if (result.type !== 'setup_status' && result.type !== 'stream_event') return;
    }
  } finally {
    sock.destroy();
  }
}

// ---------------------------------------------------------------------------
// Activity emission
// ---------------------------------------------------------------------------

/**
 * Emit a Primordial Protocol activity event to stdout.
 * Call this to let the parent agent / TUI see sub-agent progress.
 */
export function emitActivity(tool, description, messageId) {
  const event = { type: 'activity', tool, description };
  if (messageId) event.message_id = messageId;
  process.stdout.write(JSON.stringify(event) + '\n');
}

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Search for agents by capability.
 * @param {string} query - Natural language description (e.g. "web research").
 * @returns {Promise<Array>} List of agent objects with name, description, url, stars.
 */
export async function search(query) {
  const result = await request({ type: 'search', query });
  return result?.agents || [];
}

/**
 * List all available agents sorted by popularity.
 * @returns {Promise<Array>} List of agent objects.
 */
export async function searchAll() {
  const result = await request({ type: 'search_all' });
  return result?.agents || [];
}

/**
 * Spawn a sub-agent in its own sandbox.
 * @param {string} agentUrl - GitHub URL of the agent to spawn.
 * @param {Object} [options]
 * @param {Function} [options.onStatus] - Callback(event) for setup status updates.
 * @returns {Promise<string>} session_id
 */
export async function runAgent(agentUrl, { onStatus } = {}) {
  for await (const event of requestStream({ type: 'run', agent_url: agentUrl })) {
    if (event.type === 'setup_status') {
      if (onStatus) onStatus(event);
    } else if (event.type === 'session') {
      return event.session_id;
    } else if (event.type === 'error') {
      throw new Error(event.error || 'Failed to start agent');
    }
  }
  throw new Error('Unexpected end of stream during agent startup');
}

/**
 * Send a message to a sub-agent and wait for the response.
 * @param {string} sessionId - Session ID from runAgent.
 * @param {string} content - Message text to send.
 * @param {Object} [options]
 * @param {Function} [options.onActivity] - Callback(tool, description) for activity events.
 * @returns {Promise<Object>} { response: string, activities: Array }
 */
export async function messageAgent(sessionId, content, { onActivity } = {}) {
  const activities = [];
  let response = '';

  for await (const event of messageAgentStream(sessionId, content)) {
    if (event.type !== 'stream_event') continue;
    const inner = event.event || {};
    if (inner.type === 'activity') {
      const tool = inner.tool || '';
      const desc = inner.description || '';
      activities.push({ tool, description: desc });
      if (onActivity) onActivity(tool, desc);
    } else if (inner.type === 'response' && inner.done) {
      response = inner.content || '';
    }
  }

  return { response, activities };
}

/**
 * Send a message and yield raw stream events.
 * @param {string} sessionId - Session ID from runAgent.
 * @param {string} content - Message text to send.
 * @yields {Object} Raw event dicts from the delegation proxy.
 */
export async function* messageAgentStream(sessionId, content) {
  yield* requestStream({
    type: 'message',
    session_id: sessionId,
    content,
  });
}

/**
 * View the last 1000 lines of a sub-agent's output.
 * @param {string} sessionId - Session ID from runAgent.
 * @returns {Promise<Array<string>>} Output lines.
 */
export async function monitorAgent(sessionId) {
  const result = await request({ type: 'monitor', session_id: sessionId });
  return result?.lines || [];
}

/**
 * Shut down a sub-agent and release its sandbox.
 * @param {string} sessionId - Session ID from runAgent.
 */
export async function stopAgent(sessionId) {
  await request({ type: 'stop', session_id: sessionId });
}

File: packages/client/src/primordial/sandbox/primordial_delegate.py
"""Primordial Delegation SDK for Python.

Provides a simple API for agents to search, spawn, and interact with other
agents on the Primordial AgentStore via the delegation socket.

Usage:
    from primordial_delegate import search, run_agent, message_agent, stop_agent

STDLIB ONLY — no third-party dependencies.
"""

import json
import socket
import sys

SOCK_PATH = "/tmp/_primordial_delegate.sock"


# ---------------------------------------------------------------------------
# Socket helpers (fresh connection per call)
# ---------------------------------------------------------------------------

def _connect():
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    sock.connect(SOCK_PATH)
    return sock


def _send(sock, obj):
    sock.sendall((json.dumps(obj) + "\n").encode())


def _read_line(sock, buf=b""):
    while b"\n" not in buf:
        chunk = sock.recv(8192)
        if not chunk:
            raise ConnectionError("Delegation socket closed")
        buf += chunk
    line, buf = buf.split(b"\n", 1)
    return json.loads(line), buf


def _request(msg):
    """Send a command and return a single response."""
    sock = _connect()
    try:
        _send(sock, msg)
        result, _ = _read_line(sock)
        if result.get("type") == "error":
            raise RuntimeError(result.get("error", "unknown error"))
        return result
    finally:
        sock.close()


def _request_stream(msg):
    """Send a command and yield responses until stream ends."""
    sock = _connect()
    buf = b""
    try:
        _send(sock, msg)
        while True:
            result, buf = _read_line(sock, buf)
            yield result
            if result.get("type") == "error":
                return
            if result.get("done", False):
                return
            # For run: stream ends on non-setup_status
            if result.get("type") not in ("setup_status", "stream_event"):
                return
    finally:
        sock.close()


# ---------------------------------------------------------------------------
# Activity emission
# ---------------------------------------------------------------------------

def emit_activity(tool, description, message_id=None):
    """Emit a Primordial Protocol activity event to stdout.

    Call this to let the parent agent / TUI see sub-agent progress
    in real-time.
    """
    event = {"type": "activity", "tool": tool, "description": description}
    if message_id:
        event["message_id"] = message_id
    sys.stdout.write(json.dumps(event) + "\n")
    sys.stdout.flush()


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

def search(query):
    """Search for agents by capability.

    Args:
        query: Natural language description (e.g. "web research").

    Returns:
        List of agent dicts with name, description, url, stars.
    """
    result = _request({"type": "search", "query": query})
    return result.get("agents", [])


def search_all():
    """List all available agents sorted by popularity.

    Returns:
        List of agent dicts with name, description, url, stars.
    """
    result = _request({"type": "search_all"})
    return result.get("agents", [])


def run_agent(agent_url, on_status=None):
    """Spawn a sub-agent in its own sandbox.

    Args:
        agent_url: GitHub URL of the agent to spawn.
        on_status: Optional callback(event_dict) called for each setup status.

    Returns:
        session_id string.

    Raises:
        RuntimeError: If the agent fails to start.
    """
    for event in _request_stream({"type": "run", "agent_url": agent_url}):
        if event.get("type") == "setup_status":
            if on_status:
                on_status(event)
        elif event.get("type") == "session":
            return event["session_id"]
        elif event.get("type") == "error":
            raise RuntimeError(event.get("error", "Failed to start agent"))
    raise RuntimeError("Unexpected end of stream during agent startup")


def message_agent(session_id, content, on_activity=None):
    """Send a message to a sub-agent and wait for the response.

    Args:
        session_id: Session ID from run_agent.
        content: Message text to send.
        on_activity: Optional callback(tool, description) for activity events.

    Returns:
        Dict with "response" (str) and "activities" (list of dicts).
    """
    activities = []
    response = ""

    for event in message_agent_stream(session_id, content):
        if event.get("type") != "stream_event":
            continue
        inner = event.get("event", {})
        if inner.get("type") == "activity":
            tool = inner.get("tool", "")
            desc = inner.get("description", "")
            activities.append({"tool": tool, "description": desc})
            if on_activity:
                on_activity(tool, desc)
        elif inner.get("type") == "response" and inner.get("done"):
            response = inner.get("content", "")

    return {"response": response, "activities": activities}


def message_agent_stream(session_id, content):
    """Send a message and yield raw stream events.

    Args:
        session_id: Session ID from run_agent.
        content: Message text to send.

    Yields:
        Raw event dicts from the delegation proxy.
    """
    yield from _request_stream({
        "type": "message",
        "session_id": session_id,
        "content": content,
    })


def monitor_agent(session_id):
    """View the last 1000 lines of a sub-agent's output.

    Args:
        session_id: Session ID from run_agent.

    Returns:
        List of output line strings.
    """
    result = _request({"type": "monitor", "session_id": session_id})
    return result.get("lines", [])


def stop_agent(session_id):
    """Shut down a sub-agent and release its sandbox.

    Args:
        session_id: Session ID from run_agent.

    Raises:
        RuntimeError: If the session is unknown.
    """
    _request({"type": "stop", "session_id": session_id})
